src/puya/client_gen.py

```diff
@@ -0,0 +1,196 @@
+import argparse
+import json
+import typing
+from collections.abc import Iterable, Mapping, Sequence
+from pathlib import Path
+
+import attrs
+import structlog
+from immutabledict import immutabledict
+
+from puya.arc32 import OCA_ARC32_MAPPING, write_arc32_client
+from puya.errors import PuyaError
+from puya.logging_config import LogLevel, configure_logging
+from puya.models import (
+    ARC4Method,
+    ARC4MethodArg,
+    ARC4MethodConfig,
+    ARC4Returns,
+    ARC32StructDef,
+    OnCompletionAction,
+)
+
+logger = structlog.get_logger(__file__)
+ARC32_OCA_MAPPING = {v: k for k, v in OCA_ARC32_MAPPING.items()}
+
+
+@attrs.define(kw_only=True)
+class PuyaGenOptions:
+    paths: Sequence[Path] = attrs.field(default=(), repr=lambda p: str(list(map(str, p))))
+    log_level: LogLevel = LogLevel.info
+
+
+def main() -> None:
+    parser = argparse.ArgumentParser(
+        prog="puya-gen",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+        description="Output puyapy contract client for typed ARC4 ABI calls from an "
+        "ARC32 application.json spec",
+    )
+    parser.add_argument("paths", type=Path, nargs="+", metavar="PATH")
+    options = PuyaGenOptions()
+    parser.parse_args(namespace=options)
+    configure_logging(min_log_level=options.log_level)
+    output_stubs(options.paths)
+
+
+def output_stubs(paths: Sequence[Path]) -> None:
+    try:
+        app_spec_paths = resolve_app_specs(paths)
+        for app_spec_path in app_spec_paths:
+            name, methods = _parse_app_spec_methods(app_spec_path)
+            write_arc32_client(name, methods, app_spec_path.parent)
+    except PuyaError as ex:
+        logger.error(str(ex))
+
+
+def resolve_app_specs(paths: Sequence[Path]) -> Sequence[Path]:
+    app_specs = list[Path]()
+    for path in paths:
+        if not path.is_dir():
+            app_specs.append(path)
+        else:
+            patterns = ["application.json", "*.arc32.json"]
+
+            app_specs = []
+            for pattern in patterns:
+                app_specs.extend(path.rglob(pattern))
+
+    app_specs = list(set(app_specs))
+    app_specs.sort()
```

Comment:
`sorted` returns a `list`, so this could just be `app_specs = sorted(set(app_specs))`